spanner/oc_test.go

```diff
@@ -16,42 +16,217 @@ package spanner
+		}
+		row := stat.Rows[0]
+		m := getTagMap(row.Tags)
+		checkCommonTags(t, m)
+		if got, want := fmt.Sprintf("%v", row.Data), value; got != want {
+			t.Fatalf("Incorrect data: got %v, want %v", got, want)
+		}
+	case <-time.After(1 * time.Second):
+		t.Fatal("no stats were exported before timeout")
+	}
+}
+
+func TestOCStats_SessionPool_SessionsCount(t *testing.T) {
+	te := testutil.NewTestExporter(SessionsCountView)
+	defer te.Unregister()
+
+	_, client, teardown := setupMockedTestServerWithConfig(t, ClientConfig{SessionPoolConfig: DefaultSessionPoolConfig})
+	defer teardown()
+
+	client.Single().ReadRow(context.Background(), "Users", Key{"alice"}, []string{"email"})
+
+	// Wait for a while to see all exported metrics.
+	time.Sleep(100 * time.Millisecond)
+
+	// Wait until we see data from the view.
+	select {
+	case stat := <-te.Stats:
+		// There are 4 types for this metric, so we should see at least four
+		// rows.
+		if len(stat.Rows) < 4 {
+			t.Fatal("No enough metrics are exported")
+		}
+		if got, want := stat.View.Measure.Name(), statsPrefix+"num_sessions_in_pool"; got != want {
+			t.Fatalf("Incorrect measure: got %v, want %v", got, want)
+		}
+		for _, row := range stat.Rows {
+			m := getTagMap(row.Tags)
+			checkCommonTags(t, m)
+			// view.AggregationData does not have a way to extract the value. So
+			// we have to convert it to a string and then compare with expected
+			// values.
+			got := fmt.Sprintf("%v", row.Data)
+			var want string
+			switch m[tagKeyType] {
+			case "num_write_prepared_sessions":
+				want = "&{20}"
+			case "num_read_sessions":
+				want = "&{80}"
+			case "num_sessions_being_prepared":
+				want = "&{0}"
+			case "num_in_use_sessions":
+				want = "&{0}"
+			default:
+				t.Fatalf("Incorrect type: %v", m[tagKeyType])
+			}
+			if got != want {
+				t.Fatalf("Incorrect data: got %v, want %v", got, want)
+			}
+		}
+	case <-time.After(1 * time.Second):
+		t.Fatal("no stats were exported before timeout")
+	}
+}
+
+func TestOCStats_SessionPool_GetSessionTimeoutsCount(t *testing.T) {
+	te := testutil.NewTestExporter(GetSessionTimeoutsCountView)
+	defer te.Unregister()
+
+	_, client, teardown := setupMockedTestServer(t)
+	defer teardown()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Millisecond)
+	defer cancel()
+	client.Single().ReadRow(ctx, "Users", Key{"alice"}, []string{"email"})
+
+	// Wait for a while to see all exported metrics.
+	time.Sleep(100 * time.Millisecond)
+
+	// Wait until we see data from the view.
+	select {
+	case stat := <-te.Stats:
+		if len(stat.Rows) == 0 {
+			t.Fatal("No metrics are exported")
+		}
+		if got, want := stat.View.Measure.Name(), statsPrefix+"get_session_timeouts"; got != want {
+			t.Fatalf("Incorrect measure: got %v, want %v", got, want)
+		}
+		row := stat.Rows[0]
+		m := getTagMap(row.Tags)
+		checkCommonTags(t, m)
+		if got, want := fmt.Sprintf("%v", row.Data), "&{1}"; got != want {
```

Comment:
While it is highly likely that this is true, I'm a little bit afraid that this check might turn out to be flaky. The test uses a context with a timeout of 1ms, which means that it is highly likely to timeout, but there is no real guarantee that it will happen while waiting for a session. I think that adding a simulated execution time of 2ms for the `BatchCreateSessions` on the mock server would guarantee it.

Reply:
Thanks for the suggestion. Done.